python/grass/jupyter/timeseries.py

```diff
@@ -0,0 +1,435 @@
+            raise NameError(
+                _(
+                    "Could not find space time raster or vector " "dataset named {}"
+                ).format(timeseries)
+            )
+
+        # Create a temporary directory for our PNG images
+        # Resource managed by weakref.finalize.
+        self._tmpdir = (
+            # pylint: disable=consider-using-with
+            tempfile.TemporaryDirectory()
+        )
+
+        def cleanup(tmpdir):
+            tmpdir.cleanup()
+
+        weakref.finalize(self, cleanup, self._tmpdir)
+
+        # Handle Regions
+        region_manager = RegionManagerForTimeSeries(use_region, saved_region, self._env)
+        region_manager.set_region_from_timeseries(self.timeseries)
+
+    @property
+    def overlay(self):
+        """Add overlay to TimeSeries visualization"""
+        self._layers_rendered = False
+        return self._overlays
+
+    @property
+    def baselayer(self):
+        """Add base layer to TimeSeries visualization"""
+        self._layers_rendered = False
+        return self._baselayers
+
+    def set_background_color(self, color):
+        """Set background color of images.
+
+        Passed to d.rast and d.erase. Either a standard color name, R:G:B triplet, or
+        Hex. Default is White. (add example with hex)"""
+        self._bgcolor = color
+        self._layers_rendered = False
+
+    def d_legend(self, **kwargs):
+        """Display legend.
+
+        Wraps d.legend and uses same keyword arguments.
+        """
+        self._legend = kwargs
+        # If d_legend has been called, we need to re-render layers
+        self._layers_rendered = False
+
+    def _render_legend(self, img):
+        """Add legend to GrassRenderer instance"""
+        info = gs.parse_command(
+            "t.info", input=self.timeseries, flags="g", env=self._env
+        )
+        min_min = info["min_min"]
+        max_max = info["max_max"]
+        img.d_legend(
+            raster=self._layers[0],
+            range=f"{min_min}, {max_max}",
+            **self._legend,
+        )
+
+    def _render_blank_layer(self, filename):
+        """Write blank image for gaps in time series.
+
+        Adds overlays and legend to base map.
+        """
+        img = GrassRenderer(
+            filename=filename, use_region=True, env=self._env, read_file=True
+        )
+        for grass_module, kwargs in self._overlays.calls:
+            img.run(grass_module, **kwargs)
+        # Add legend if needed
+        if self._legend:
+            self._render_legend(img)
+
+    def render(self):
+        """Renders image for each time-step in space-time dataset.
+
+        Save PNGs to temporary directory. Must be run before creating a visualization
+        (i.e. time_slider or animate). Can be time-consuming to run with large
+        space-time datasets.
+        """
+
+        # Make base image (background and baselayers)
+        random_name_base = gs.append_random("base", 8) + ".png"
```

Comment:
I forgot why this has a randomized name. Perhaps a comment explaining it if it is needed.

Maybe not a big deal, but given that the old files are still around perhaps delete them or overwrite them.

Reply:
Before we change to random, the file was called "base.png". This would create an issue if there was a layer called "base" (which would also create a file called "base.png").

Reply:
I see. Can you please add this to a comment, so it is clear why you are creating a random, sort of unique name?

A comment on how it overwrites existing files when it runs render again would be helpful too. (To answer, e.g., how does it ensure that old file content is not read when new rendering starts?)